crates/oxc_transformer/src/es2018/object_spread.rs

```diff
@@ -0,0 +1,139 @@
+//! ES2018 object spread transformation.
+//!
+//! This plugin transforms object spread properties (`{ ...x }`) to a series of `_objectSpread` calls.
+//!
+//! > This plugin is included in `preset-env`, in ES2018
+//!
+//! ## Example
+//!
+//! Input:
+//! ```js
+//! var x = { a: 1, b: 2 };
+//! var y = { ...x, c: 3 };
+//! ```
+//!
+//! Output:
+//! ```js
+//! var x = { a: 1, b: 2 };
+//! var y = _objectSpread({}, x, { c: 3 });
+//! ```
+//!
+//! ## Implementation
+//!
+//! Implementation based on [@babel/plugin-transform-object-rest-spread](https://babeljs.io/docs/babel-plugin-transform-object-rest-spread).
+//!
+//! ## References:
+//! * Babel plugin implementation: <https://github.com/babel/babel/tree/main/packages/babel-plugin-transform-object-rest-spread>
+//! * Object rest/spread TC39 proposal: <https://github.com/tc39/proposal-object-rest-spread>
+
+use super::object_rest_spread::ObjectRestSpreadOptions;
+use crate::context::Ctx;
+
+use oxc_ast::ast::*;
+use oxc_semantic::ReferenceFlags;
+use oxc_span::SPAN;
+use oxc_traverse::{Traverse, TraverseCtx};
+
+pub struct ObjectSpread<'a> {
+    _ctx: Ctx<'a>,
+    options: ObjectRestSpreadOptions,
+}
+
+impl<'a> ObjectSpread<'a> {
+    pub fn new(options: ObjectRestSpreadOptions, ctx: Ctx<'a>) -> Self {
+        Self { _ctx: ctx, options }
+    }
+}
+impl<'a> Traverse<'a> for ObjectSpread<'a> {
+    fn enter_expression(&mut self, expr: &mut Expression<'a>, ctx: &mut TraverseCtx<'a>) {
+        let Expression::ObjectExpression(obj_expr) = expr else {
+            return;
+        };
+
+        if obj_expr
+            .properties
+            .iter()
+            .all(|prop| matches!(prop, ObjectPropertyKind::ObjectProperty(..)))
+        {
+            return;
+        }
```

Comment:
I generally believe the performance here is acceptable, as it operates at O(n) complexity and will early return.
However, if we delegate the detection logic to another visitor, we'd have to manage the interdependencies between them.
Additionally, there's a risk that once `exit_spread_element` is exited, it could be mutated by another padd visitor, leading to the flag being out of sync with the actual state.

Reply:
All of that is true - co-ordinating multiple visitors is tricky. But please note my comment above that the addition of this transform is slowing down the transformer on one benchmark by 28%. That's not negligible.

I may be wrong that these lines is where the perf loss is coming from, but it's my best guess.

Reply:
I removed the `self.enter_expression(expr, ctx);` above and the CodSpeed gate passed. So I think the bottleneck is there, not here.

Reply:
The benchmark of pdf.mjs has returned to its original level.

<img width="848" alt="image" src="https://github.com/user-attachments/assets/e41a1191-9b36-44a7-89e8-4b2824dd5937">

Reply:
It's really odd because `pdf.js` only has 4 x object spreads in it. So I can't explain how that could be enough to produce a 28% perf drop. But it seems it did. Great!